File: UnityProject/Assets/Scripts/Core/MissionTracker.cs
using System.Collections.Generic;
using System.Linq;

namespace MiniMarketShift.Core
{
    internal sealed class MissionTracker
    {
        private readonly List<MissionRuntime> _missions;

        public MissionTracker(IEnumerable<MissionDefinition> missions)
        {
            _missions = missions.Select(m => new MissionRuntime(m)).ToList();
        }

        public void RegisterAction(string type, int amount)
        {
            foreach (var mission in _missions)
            {
                if (mission.Type == type)
                {
                    mission.Progress += amount;
                }
            }
        }

        public List<MissionStatus> GetStatus()
        {
            return _missions.Select(m => new MissionStatus
            {
                id = m.Id,
                name = m.Name,
                type = m.Type,
                progress = m.Progress,
                goal = m.Goal,
                reward_coins = m.RewardCoins,
                completed = m.Progress >= m.Goal,
                claimed = m.Claimed
            }).ToList();
        }

        public int Claim(string missionId)
        {
            var mission = _missions.FirstOrDefault(m => m.Id == missionId);
            if (mission == null || mission.Claimed || mission.Progress < mission.Goal)
            {
                return 0;
            }

            mission.Claimed = true;
            return mission.RewardCoins;
        }

        private sealed class MissionRuntime
        {
            public string Id { get; }
            public string Name { get; }
            public string Type { get; }
            public int Goal { get; }
            public int RewardCoins { get; }
            public int Progress { get; set; }
            public bool Claimed { get; set; }

            public MissionRuntime(MissionDefinition definition)
            {
                Id = definition.id;
                Name = definition.name;
                Type = definition.type;
                Goal = definition.goal;
                RewardCoins = definition.reward_coins;
                Progress = 0;
                Claimed = false;
            }
        }
    }
}

File: src/domain/mission-tracker.mjs
export class MissionTracker {
  constructor(missions) {
    this.missions = missions.map((mission) => ({
      ...mission,
      progress: 0,
      claimed: false
    }));
  }

  registerAction(type, amount = 1) {
    for (const mission of this.missions) {
      if (mission.type === type) {
        mission.progress += amount;
      }
    }
  }

  getStatus() {
    return this.missions.map((mission) => ({
      id: mission.id,
      name: mission.name,
      type: mission.type,
      progress: mission.progress,
      goal: mission.goal,
      reward_coins: mission.reward_coins,
      completed: mission.progress >= mission.goal,
      claimed: mission.claimed
    }));
  }

  claim(missionId) {
    const mission = this.missions.find((item) => item.id === missionId);
    if (!mission || mission.claimed || mission.progress < mission.goal) {
      return 0;
    }

    mission.claimed = true;
    return mission.reward_coins;
  }
}
